fix(quota): handle undefined userTier on initial render

sheet00 opened the pull request:
Fixes #3805

When a paid-tier user exceeds their quota, they were incorrectly shown the message for free-tier users.

This was caused by a timing issue in the `useQuotaAndFallback` hook, where the `userTier` prop is `undefined` during the initial component render. The fallback handler would capture this `undefined` value and incorrectly identify the user as being on the free tier.

This commit fixes the issue by using the nullish coalescing operator to fall back to `config.getUserTier()` when the `userTier` prop is undefined. This ensures the correct tier information is used even before the React state has fully propagated.

A regression test is added to verify that the correct (paid-tier) message is displayed when the `userTier` prop is undefined.

## TLDR

Fixes a bug where paid-tier users saw the free-tier quota message by adding a fallback to get the correct user tier during the component's initial render phase.

## Dive Deeper

The root cause was a race condition during app startup. The `useQuotaAndFallback` hook would initialize with an `undefined` `userTier` prop, creating a `fallbackHandler` with a stale closure. If a quota error occurred before the `userTier` prop was updated with the correct value post-authentication, this stale handler would treat a paid user as a free user.

The fix makes the handler robust by checking `config.getUserTier()` at the moment the error occurs, ensuring the most up-to-date information is always used.

<details>
<summary>Detailed Investigation Report</summary>

# Final Investigation Report: Paid Tier Quota Message Issue

## 1. Phenomenon

When a paid-tier user exceeds their quota, they are shown the "upgrade to a Gemini Code Assist Standard or Enterprise plan" message intended for free-tier users, instead of the correct "switch to using a paid API key from AI Studio" message.

## 2. Root Cause

A timing issue in the React component lifecycle.
- **Flow of the problem**:
  1. On initial render of the `AppContainer.tsx` component, the `userTier` state is initialized as `undefined`.
  2. The `useQuotaAndFallback` hook is called with `userTier=undefined`, and its `useEffect` runs.
  3. The `fallbackHandler` generated within this `useEffect` captures the stale information that `userTier` is `undefined` in its closure.
  4. If a quota error occurs before authentication is complete and the `userTier` state is updated with the correct value, this "stale handler" is invoked.
  5. The `isPaidTier` check is performed on `undefined`, resulting in `false`, causing the paid user to be treated as a free user.

- **Problem location**: `packages/cli/src/ui/hooks/useQuotaAndFallback.ts`

## 3. Evidence in Production Code

The hypothesis that the `userTier` property could be `undefined` on initialization was confirmed by investigating the production code in `packages/cli/src/ui/AppContainer.tsx`.

- **`userTier` Initialization**:
  The `userTier` is initialized as `undefined` as component state.
  ```typescript
  // packages/cli/src/ui/AppContainer.tsx:190
  const [userTier, setUserTier] = useState<UserTierId | undefined>(undefined);
  ```

- **`userTier` Update**:
  The actual value is set later within a `useEffect` hook when the authentication state changes.
  ```typescript
  // packages/cli/src/ui/AppContainer.tsx:368-372
  useEffect(() => {
    if (authState === AuthState.Authenticated) {
      setUserTier(config.getUserTier());
    }
  }, [config, authState]);
  ```
This confirms that a window exists between the initial mount and the completion of authentication where `userTier` is `undefined`, and an error during this window will trigger the issue.

## 4. Resolution Process

The fix and verification were carried out with the following steps.

### Step 1: Add a Regression Test

First, a test case to reproduce this bug was added to `packages/cli/src/ui/hooks/useQuotaAndFallback.test.ts`. This test verifies that the correct message for paid users is displayed when the `userTier` property is `undefined`.

```typescript
// The getRegisteredHandler helper has a default value for userTier,
// so we call renderHook directly to ensure we are testing the
// truly `undefined` case.
it('should use config.getUserTier as fallback when userTier prop is undefined', async () => {
  // Arrange
  vi.spyOn(mockConfig, 'getUserTier').mockReturnValue(UserTierId.STANDARD);
  mockedIsGenericQuotaExceededError.mockReturnValue(true);

  renderHook(() =>
    useQuotaAndFallback({
      config: mockConfig,
      historyManager: mockHistoryManager,
      userTier: undefined,
      setAuthState: mockSetAuthState,
      setModelSwitchedFromQuotaError: mockSetModelSwitchedFromQuotaError,
    }),
  );
  const handler = setFallbackHandlerSpy.mock.calls[0][0] as FallbackModelHandler;

  // Act
  const result = await handler('model-A', 'model-B', new Error('quota exceeded'));

  // Assert
  expect(result).toBe('stop');
  const message = (mockHistoryManager.addItem as Mock).mock.calls[0][0].text;
  expect(message).toContain('switch to using a paid API key from AI Studio');
  expect(message).not.toContain('upgrade to a Gemini Code Assist Standard or Enterprise plan');
});
```
Running this test with the code before the fix confirmed that the issue was reproduced (the free-tier message was displayed instead of the paid-tier message).

### Step 2: Implement the Fix

Next, to resolve this issue, `packages/cli/src/ui/hooks/useQuotaAndFallback.ts` was modified. A fallback to get the value from `config.getUserTier()` was added for when `userTier` is `undefined`.

```typescript
// packages/cli/src/ui/hooks/useQuotaAndFallback.ts
// The userTier prop can be undefined on initial render, so we fall back to the config value.
const currentUserTier = userTier ?? config.getUserTier();
const isPaidTier =
  currentUserTier === UserTierId.LEGACY ||
  currentUserTier === UserTierId.STANDARD;
```

### Step 3: Verify the Fix

After applying the fix, the tests were run again, and it was confirmed that all tests, including the added test case, passed. This ensures that the fix is correct and no regressions have occurred.

## 5. Final Conclusion

A race condition caused by React's initialization timing was identified, and a robust fix was implemented. Specifically, logic was added to fall back to the config when the `userTier` property is `undefined`, and a regression test was also added to guarantee this behavior. With this, the issue has been permanently resolved.

</details>

## Reviewer Test Plan

1.  The primary validation is the new automated test case added in this PR (`should use config.getUserTier as fallback when userTier prop is undefined`). A reviewer can confirm the fix by running the test and seeing it pass:
    ```bash
    npm run test --workspace @google/gemini-cli -- src/ui/hooks/useQuotaAndFallback.test.ts
    ```
2.  Manual verification is difficult as it requires triggering a quota error in the narrow time window between app startup and user authentication completing. The automated test is the most reliable method to confirm the fix.

## Testing Matrix

|          | 🍏  | 🪟  | 🐧  |
| -------- | --- | --- | --- |
| npm run  | ❓  | ❓  | ✅  |
| npx      | ❓  | ❓  | ❓  |
| Docker   | ❓  | ❓  | ❓  |
| Podman   | ❓  | -   | -   |
| Seatbelt | ❓  | -   | -   |

## Linked issues / bugs

Fixes #3805

abhipatel12 reviewed (commented):
Hey @sheet00! Sorry again for the delayed review here. This is great work. 

I left a couple comments, so let me know what you think.

Also, I noticed the CLA check is failing. Have you completed the CLA at: https://cla.developers.google.com/

abhipatel12 commented on packages/core/src/core/loggingContentGenerator.test.ts:
```diff
@@ -0,0 +1,56 @@
+      expect(loggingGenerator.userTier).toBeUndefined();
+    });
+  });
+});
```
`npm run preflight` should also run the lint check, type check, and the unit tests. This way you can be sure that everything passes!

sheet00 commented:
hi! @abhipatel12 

Thank you for the review!
I have addressed the feedback points.
The modifications are as described in the comments above.
Please let me know if there's anything else that needs to be addressed!
I have also completed the CLA registration!

sheet00 commented:
@abhipatel12 
Hi, how did the e2e tests go? Are we ready to merge this?
